### ISSUE TITLE
SEARCH command error: BAD

### ISSUE THREAD

**Issue by mjs** [mjs/imapclient#205]
Originally reported by: **Paul Hammant (Bitbucket: [paul_hammant](https://bitbucket.org/paul_hammant))**

---

In release 1.0.1:

```
#!python

File "/usr/local/lib/python2.7/site-packages/imapclient/imapclient.py", line 725, in search
    return self._search(criteria, charset)
  File "/usr/local/lib/python2.7/site-packages/imapclient/imapclient.py", line 749, in _search
    data = self._raw_command_untagged(b'SEARCH', args)
  File "/usr/local/lib/python2.7/site-packages/imapclient/imapclient.py", line 1083, in _raw_command_untagged
    typ, data = self._raw_command(command, args)
  File "/usr/local/lib/python2.7/site-packages/imapclient/imapclient.py", line 1146, in _raw_command
    return self._imap._command_complete(to_unicode(command), tag)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/imaplib.py", line 918, in _command_complete
    raise self.error('%s command error: %s %s' % (name, typ, data))
imaplib.error: SEARCH command error: BAD ['The specified message set is invalid.']
```

I followed some guy's advice on StackOverflow - http://stackoverflow.com/a/35046977/832725 - and I'm fine with release 0.13 of imapclient.

Same code, same IMAP folder, same message for both the 1.0.1 and 0.13 invocations.
Other pertinent info - Python 2.7.11 (Mac, Homebrew)

---
- Bitbucket: https://bitbucket.org/mjs0/imapclient/issue/208

**Comment by paul-hammant** [comment]
Yup, linking to docs from exceptions works for the likes of AngularJS and Selenium2+

**Comment by mlorant** [comment]
@paul-hammant @mjs There is no complete documentation of the SEARCH arguments on imapclient (neither in imaplib AFAIK), pointing to the RFC may be a better alternative? Or pointing to imapclient doc, [which itself points to the RFC](https://imapclient.readthedocs.io/en/stable/#imapclient.IMAPClient.search)...

Does something like would look fine?
```
$ python main.py 
Traceback (most recent call last):
  File "main.py", line 27, in <module>
    test_bad_search()
  File "main.py", line 23, in test_bad_search
    conn.search('TOO wonder@world.com')
  File "../imapclient/imapclient/imapclient.py", line 730, in search
    return self._search(criteria, charset)
  File "../imapclient/imapclient/imapclient.py", line 765, in _search
    criteria=criteria, original_msg=m.group(1)
imaplib.error: Criteria "TOO wonder@world.com" is not a valid SEARCH argument. Please refer to the RFC 3501 for more information about the command syntax: https://tools.ietf.org/html/rfc3501.html#section-6.4.4
Original imaplib exception message was: 'Error in IMAP command UID SEARCH: Unknown argument TOO'
```

The idea behind is just wrapping the IMAP command with a `try/exccept` and check if the exception message starts with `SEARCH command error: BAD`. [The implementation is pretty straightforward](https://github.com/mlorant/imapclient/commit/9fdc699594f7081bf4cece7bdc7ec02a43fabb1c).

**Comment by mjs** [comment]
This is great - thanks!

I'd prefer that the error links to the IMAPClient docs, which then link to RFC. As mentioned earlier in this ticket, a common problem is that people do `['NOT DELETED']`, instead of `['NOT', 'DELETED']`. This is more of a "using IMAPClient" thing than "understanding the RFC" thing, so directing people to the docs seems more helpful. This also provides an opportunity for us to provide docs that are easier for people to digest than the RFC.

**Comment by mlorant** [comment]
Made another round with doc URL and tests cases: [commit amended](https://github.com/mlorant/imapclient/commit/7059956f213d32aa59271468fa37ac4de2228233). I could propose a PR, but one specific case is bothering me.

There is no way to make the difference between a SEARCH invalid argument and another imaplib problem. [Looking in imaplib code](https://github.com/python/cpython/blob/3.6/Lib/imaplib.py#L1009), you will see that any `BAD` command response is treated the same way and there is no indication in the RFC about how the exception message should be formatted. 
This may happen for example if no mailbox is selected yet. To "handle" this, I've changed the wording of the exception so it does:

```
# First execution select_folder('INBOX') / search('dummy thing')
$ python main.py 
Traceback (most recent call last):
  File "main.py", line 27, in <module>
    test_bad_search()
  File "main.py", line 23, in test_bad_search
    conn.search(['TOO', 'wonder@world.com'])
  File "../imapclient/imapclient/imapclient.py", line 730, in search
    return self._search(criteria, charset)
  File "../imapclient/imapclient/imapclient.py", line 768, in _search
    criteria='"%s"' % criteria if not isinstance(criteria, list) else criteria
imaplib.error: 'Error in IMAP command UID SEARCH: Unknown argument TOO'
This error may have been caused by a syntax error in the criteria ['TOO', 'wonder@world.com'].
Please refer to the documentation for more information about the command syntax: https://imapclient.readthedocs.io/en/stable/#imapclient.IMAPClient.search

# Second execution, without the select_folder
$ python main.py 
Traceback (most recent call last):
  File "main.py", line 27, in <module>
    test_bad_search()
  File "main.py", line 23, in test_bad_search
    conn.search(['TOO', 'wonder@world.com'])
  File "../imapclient/imapclient/imapclient.py", line 730, in search
    return self._search(criteria, charset)
  File "../imapclient/imapclient/imapclient.py", line 768, in _search
    criteria='"%s"' % criteria if not isinstance(criteria, list) else criteria
imaplib.error: 'No mailbox selected.'
This error may have been caused by a syntax error in the criteria ['TOO', 'wonder@world.com'].
Please refer to the documentation for more information about the command syntax: https://imapclient.readthedocs.io/en/stable/#imapclient.IMAPClient.search
```

This is not the best I would want but knowing the internals of imaplib and the lack of message responses definition in the RFC, I don't think we can do better there.

**Comment by mjs** [comment]
Despite the limitations, this is still an improvement. 

May I suggest the following minor variation to the wording:

```
imaplib.error: 'Error in IMAP command UID SEARCH: Unknown argument TOO'

This may have been caused by an error in the criteria: ['TOO', 'wonder@world.com']
Please refer to the documentation for more information about search criteria syntax.
https://imapclient.readthedocs.io/en/stable/#imapclient.IMAPClient.search
```

The blank line helps to ensure that exception text isn't lost and the having the docs URL on a separate line makes it easier to find too.

Thank you. Please propose a PR!

**Comment by mjs** [comment]
Closed by #262.
